### ISSUE TITLE
Add `sourceBuilder` to Data.Conduit.ByteString.Builder

### ISSUE THREAD

**Issue by dylex** [snoyberg/conduit#309]
It'd be nice to have a version of a function in `Data.Conduit.ByteString.Builder` like:
```
sourceBuilder :: Builder -> Producer m ByteString
sourceBuilder = sourceList . toChunks
sourceBuilder b = yield b .| builderToByteString
```
The first is bad because laziness. The second may actually be just fine but feels a bit silly, but if this is the recommended solution, I'll withdraw the request.  Could also be called `sourceBuilderAsByteString` or something, instead, and possibly have `sourceByteStringWith` allocation strategy versions, too.

**Comment by snoyberg** [comment]
Most likely the solution you're looking for would be based on a more
generalized version of toByteStringIO:

https://www.stackage.org/haddock/lts-8.13/blaze-builder-0.4.0.2/Blaze-ByteString-Builder.html#v:toByteStringIO

However, to my knowledge such a function doesn't exist, and
builderToByteString is the closest thing to it. In theory toByteStringIO
could be more efficient since it has more knowledge than
builderToByteString, but before duplicating such low-level, error-prone
code it would be good to see that there really is a performance difference.

**Comment by dylex** [comment]
I guess this is more complicated than I thought. I can see how

https://github.com/fpco/streaming-commons/blob/master/Data/Streaming/ByteString/Builder.hs#L78 (that `builderToByteString` uses)

could be less efficient for this, though now I realize there may be cases where

http://hackage.haskell.org/package/streaming-commons-0.1.16/docs/src/Data.Streaming.ByteString.Builder.html#toByteStringIOWith

which re-uses the same buffer for each chunk (like `unsafeBufferToByteString`, if I understand it correctly) could be even better.

I guess the impediment to using either of these directly is lifting them into other monads.  It seems like it wouldn't be too hard to make them:
```
toByteStringMIO :: MonadIO m => (ByteString -> m ()) -> Builder -> m ()
```
In fact, could this even be done using `MonadTrans` as-is, and then essentially call them as `toByteStringIO yield builder`?

**Comment by snoyberg** [comment]
No, `MonadTrans` isn't sufficient, since the `m ()` appears in negative position. This would require something like `MonadBaseControl`, or—as you point out—just generalizing to `MonadIO`. However, I'd point out that even that is debatable. The operation is morally pure, so in theory this could be a place to use `unsafePerformIO` and get the much more general `Monad m` constraint. But at that point, it's just the same as `toLazyByteString`. Which does beg the question: what's the downside of the laziness you see in this case?

**Comment by dylex** [comment]
In this case happen to be in IO, but I see how making the IO-specific version of `sourceBuilder` to avoid laziness would be overly restrictive.  I suppose I'd suggest the existance of `sourceBuilder` with whichever simple implementation so anyone wondering how to turn a Builder into a ByteString Source has an easy answer that could be improved if someone ever finds reason.  But feel free to just close, too, and thanks for the insight.

**Comment by snoyberg** [comment]
I'm weakly +1 on adding a `sourceBuilder` function. Your arguments make sense, but my slight reservations are around the relative lack of added convenient (vs `sourceLazy . toLazyByteString`) and the real possibility of choosing the wrong type signature.
